File: common-local-task-message/src/main/java/com/dev/lib/local/task/message/storage/LocalTaskMessageStorage.java
package com.dev.lib.local.task.message.storage;

import com.dev.lib.local.task.message.data.TaskMessageRepository;
import com.dev.lib.local.task.message.domain.adapter.ILocalTaskMessageAdapt;
import com.dev.lib.local.task.message.domain.model.NotifyType;
import com.dev.lib.local.task.message.domain.model.entity.TaskMessageEntityCommand;
import com.dev.lib.mq.MessageExtend;
import com.dev.lib.mq.reliability.MessageStorage;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Component
public class LocalTaskMessageStorage implements MessageStorage {

    private final ILocalTaskMessageAdapt adapt;
    private final TaskMessageRepository repository;

    public LocalTaskMessageStorage(ILocalTaskMessageAdapt adapt, TaskMessageRepository repository) {
        this.adapt = adapt;
        this.repository = repository;
    }

    @Override
    public void saveBlocking(MessageExtend<?> message) {
        String destination = (String) message.getHeaders().getOrDefault("destination", "default");
        TaskMessageEntityCommand cmd = createCommand(message, destination);
        adapt.saveMessage(cmd);
    }

    @Override
    public void markAsConsumedBlocking(String messageId) {
        adapt.updateTaskStatusToSuccess(messageId);
    }

    @Override
    public List<MessageExtend<?>> getPendingMessagesBlocking(int limit) {
        List<Integer> houseNumbers = java.util.stream.IntStream.range(0, 10)
                .boxed()
                .collect(Collectors.toList());
        List<TaskMessageEntityCommand> cmdList = adapt.selectByHouseNumber(houseNumbers, "", limit);

        return cmdList.stream()
                .map(cmd -> {
                    TaskMessageEntityCommand.NotifyConfig config = cmd.getNotifyConfig();
                    if (config != null && config.getMq() != null) {
                        return (MessageExtend<?>) config.getMq().getPayload().get("message");
                    }
                    return null;
                })
                .filter(msg -> msg != null)
                .collect(Collectors.toList());
    }

    @Override
    public void deleteBlocking(String messageId) {
        TaskMessageRepository.Query query = new TaskMessageRepository.Query();
        query.setTaskId(messageId);
        repository.physicalDelete().delete(query);
    }

    public void saveAsPending(MessageExtend<?> message, String destination) {
        TaskMessageEntityCommand cmd = createCommand(message, destination);
        adapt.saveMessage(cmd);
    }

    private TaskMessageEntityCommand createCommand(MessageExtend<?> message, String destination) {
        TaskMessageEntityCommand cmd = new TaskMessageEntityCommand();
        cmd.setTaskId(message.getId().toString());
        cmd.setTaskName((String) message.getHeaders().getOrDefault("task-name", "mq-task"));
        cmd.setBusinessId(message.getId().toString());
        cmd.setMaxRetry(message.getRetry());
        cmd.setNotifyType(NotifyType.MQ);

        TaskMessageEntityCommand.NotifyConfig notifyConfig = new TaskMessageEntityCommand.NotifyConfig();
        TaskMessageEntityCommand.NotifyConfig.Mq mqConfig = new TaskMessageEntityCommand.NotifyConfig.Mq();
        mqConfig.setDestination(destination);
        mqConfig.setPayload(Map.of(
                "message", message,
                "body", message.getBody(),
                "headers", message.getHeaders(),
                "key", message.getKey()
        ));
        notifyConfig.setMq(mqConfig);
        cmd.setNotifyConfig(notifyConfig);

        return cmd;
    }
}

File: common-mq-rabbit/src/main/kotlin/com/dev/lib/rabbit/LocalTaskMessageStorage.kt
package com.dev.lib.rabbit

import com.dev.lib.local.task.message.data.LocalTaskMessagePo
import com.dev.lib.local.task.message.data.LocalTaskStatus
import com.dev.lib.local.task.message.data.LocalTaskMessagePoToTaskMessageEntityCommandMapper
import com.dev.lib.local.task.message.data.TaskMessageRepository
import com.dev.lib.local.task.message.domain.adapter.ILocalTaskMessageAdapt
import com.dev.lib.local.task.message.domain.model.NotifyType
import com.dev.lib.local.task.message.domain.model.entity.TaskMessageEntityCommand
import com.dev.lib.mq.MessageExtend
import com.dev.lib.mq.reliability.MessageStorage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Component

@Component
class LocalTaskMessageStorage(
    private val adapt: ILocalTaskMessageAdapt,
    private val repository: TaskMessageRepository,
    private val mapper: LocalTaskMessagePoToTaskMessageEntityCommandMapper
) : MessageStorage {

    override suspend fun save(message: MessageExtend<*>) = withContext(Dispatchers.IO) {
        val cmd = TaskMessageEntityCommand().apply {
            taskId = message.id.toString()
            taskName = message.headers["task-name"] ?: "mq-task"
            businessId = message.id.toString()
            maxRetry = message.retry
            notifyType = NotifyType.MQ
            notifyConfig = TaskMessageEntityCommand.NotifyConfig().apply {
                mq = TaskMessageEntityCommand.NotifyConfig.Mq().apply {
                    destination = message.headers["destination"] as? String ?: "default"
                    payload = mapOf(
                        "message" to message,
                        "body" to message.body,
                        "headers" to message.headers,
                        "key" to message.key
                    )
                }
            }
        }
        adapt.saveMessage(cmd)
    }

    override suspend fun markAsConsumed(messageId: String) = withContext(Dispatchers.IO) {
        adapt.updateTaskStatusToSuccess(messageId)
    }

    override suspend fun getPendingMessages(limit: Int): List<MessageExtend<*>> = withContext(Dispatchers.IO) {
        val houseNumbers = listOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9)
        val cmdList = adapt.selectByHouseNumber(houseNumbers, "", limit)

        cmdList.mapNotNull { cmd ->
            val payload = cmd.notifyConfig?.mq?.payload
            @Suppress("UNCHECKED_CAST")
            val message = payload?.get("message") as? MessageExtend<*>
            message
        }
    }

    override suspend fun delete(messageId: String) = withContext(Dispatchers.IO) {
        val query = TaskMessageRepository.Query()
        query.taskId = messageId
        repository.physicalDelete().delete(query)
        Unit
    }

    fun saveAsPending(message: MessageExtend<*>, destination: String) {
        val cmd = TaskMessageEntityCommand().apply {
            taskId = message.id.toString()
            taskName = message.headers["task-name"] ?: "mq-task"
            businessId = message.id.toString()
            maxRetry = message.retry
            notifyType = NotifyType.MQ
            notifyConfig = TaskMessageEntityCommand.NotifyConfig().apply {
                mq = TaskMessageEntityCommand.NotifyConfig.Mq().apply {
                    this.destination = destination
                    payload = mapOf(
                        "message" to message,
                        "body" to message.body,
                        "headers" to message.headers,
                        "key" to message.key
                    )
                }
            }
        }
        adapt.saveMessage(cmd)
    }
}
